vignettes/brms_threading.Rmd

```diff
@@ -0,0 +1,460 @@
+  have physical cores available.
+- Ensure that the data is randomly sorted such that consecutive
+  subsets of the data are roughly of the same computational effort.
+
+## Within-chain parallelization
+
+The within-chain parallelization implemented in **brms** is based on
+the `reduce_sum` facility in Stan. The basic principle `reduce_sum`
+uses is to split a large summation into arbitrary smaller partial
+sums. Due to the commutativity and associativity of the sum operation
+these smaller partial sums can be evaluated in any order and in
+parallel from one another. **brms** leverages `reduce_sum` to evaluate
+the log-likelihood of the model in parallel as for example
+
+$$
+\begin{aligned}
+l(y|\theta) &= \sum_{i=1}^N l_i(y_i| \theta) \\
+ &= \sum_{i=1}^{S_1} l_i(y_i| \theta) + \sum_{i=S_1+1}^N l_i(y_i| \theta).
+\end{aligned}
+$$
+
+As a consequence, the within-chain parallelization requires
+independent log-likelihood terms from one another which restricts its
+applicability to some degree.
+
+Furthermore, the within-chain parallelization is only applicable to
+the evaluation of the data likelihood while all other parts of the
+model will remain running serially. Thus, only a partial fraction of
+the entire Stan model will run in parallel which limits the potential
+speedup one may obtain. The theoretical speedup for a partially in
+parallel running program is described by [Amdahl‘s
+law](https://en.wikipedia.org/wiki/Amdahl%27s_law). For example, with
+90% of the computational load running in parallel one can essentially
+double the exectuion speed with 2 cores while 8 cores may only speedup
+the program by at most 5x. How large the computational cost of the
+log-likelihood is in relation to the entire model is very dependend on
+the model of the user.
+
+In practice, the speedups are even smaller than the theoertical
+speedups. This is caused by the additional overhead implied by forming
+multiple smaller sums than just one large one. For example, for each
+partial sum formed the entire parameter vector $\theta$ has to be
+copied in memory for Stan to be able to calculate the gradient of the
+log-likelihood. Hence, with more partial sums, more copying is
+necessary as opposed to evaluating just one large sum. Whether the
+additional copying is indeed relevant depends on the computational
+cost of the log-likelihood of each term and the number of
+parameters. For a model with a computationally cheap normal
+log-likelihood this effect is more important than for a model with a
+Poisson log-likelihood and for hierarchical models with many random
+effects more copying is needed than for simpler fixed effect models.
+It may therefore be necessary to form sufficiently large partial sums
+to warrant an efficient parallel execution. The size of the partial
+sums is referred to as the `grainsize` which is set to a reasonable
+default value, but for some models this tuning parameter requires some
+attention from the user for optimal performance.
```

Comment:
Perhaps it would be useful to link to [Stan's user guide recommendations for choosing the grainsize](https://mc-stan.org/docs/2_23/stan-users-guide/reduce-sum.html#reduce-sum-grainsize) or even copy them here?

Reply:
Hmm... we can link there, sure... but what I actually hope that the text is written such that people start with the default. The section is mainly there to illustrate and visualize the overhead. Maybe a sentence like "additional information can be found in the Stan's user guide"?